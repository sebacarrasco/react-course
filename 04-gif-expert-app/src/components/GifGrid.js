import React from 'react'
import {GifGridItem} from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ( { category } ) => {

    // const [images, setImages] = useState([]);

    // // // Cuando cambie un elemento del arreglo que se le pasa
    // // // al useEffect, entonces se volver치 a aejecutar getGifs
    // // // (En este caso solo se ejecutar치 solo una vez ya que
    // // // category no cambia nunca)
    // useEffect( () => {
    //     getGifs(category)
    //         .then(setImages);
    // }, [category]);

    // Lo que est치 comentado arriba es equivalente a lo siguiente
    // (usando un custom hook que por detr치s usa useEffect)
    const { data:images, loading } = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>
            {loading && <p>Loading...</p>}
            <div className="card-grid">
                    {
                    images.map(img => 
                            <GifGridItem key={img.id} {...img}/> // Estamos haciendo spread del objeto img para
                                                                // para pasar sus propiedades "sueltas"
                    )
                    }
            </div>
        </>
    )
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
};