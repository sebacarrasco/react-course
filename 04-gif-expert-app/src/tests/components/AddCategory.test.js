import React from 'react';
import {shallow} from "enzyme";
import { AddCategory } from '../../components/AddCategory';

describe('Tests for <AddCategory />', () => {

    //  Se declaran en este scope y no dentro del beforeEacah para que
    // esten accesibles desde los tests.
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories}/>); // Se define redundantemente acá para que aparezcan las autocompletaciones

    beforeEach ( () => {
        jest.clearAllMocks(); // Para reinicializar la función (que no aparezca que fue llamada por otros tests)
        wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    });
    
    test('should match the snapshot', () => {

        expect(wrapper).toMatchSnapshot();

    });
    
    test('should change the text of the textbox (input)', () => {
        const input = wrapper.find("input");
        const value = "new input";

        input.simulate("change", { target: { value } }); // El segundo parámetro es el evento
        
        // Queda pendiente esto (por alguna razón no se actualiza el valor del input)
    });

    test('should not post the info on submit when the input is void', () => {
        
        //  El segundo parámetro de simulate es el evento
        wrapper.find("form").simulate("submit", { preventDefault: () => {} });

        // Como no se pasó un input válido, la función setCategories no debió haber sido llamada
        expect(setCategories).not.toHaveBeenCalled();
    });

    test('should call setCategories and clear the input text box when the input is valid', () => {
        const input = wrapper.find("input");
        const value = "new input";

        // Se hace el cambio en el input
        input.simulate("change", { target: { value } }); // El segundo parámetro es el evento

        // Se hace el submit
        wrapper.find("form").simulate("submit", { preventDefault: () => {} });

        // Se espera que la función haya sido llamada
        expect(setCategories).toHaveBeenCalled();
        // Podría haber sido también: .toHaveBeenCalledTimes(1);

        // Además, se espera que la función haya sido llamada con una función
        // (La función que actualizará el valor de setCategories, ej: (cats) => [...cats, new_cat])
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        // Se espera que el texto de la caja de texto sea ''
        const { value:inputValue } = input.props();
        expect(inputValue).toBe('');
    });
    
    
});
