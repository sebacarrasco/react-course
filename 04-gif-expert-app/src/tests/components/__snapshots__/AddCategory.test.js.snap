// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <AddCategory /> should match the snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
